### PR TITLE
implement new CUDA compiler checks

### PR DESCRIPTION
**Pull request by davebayer**
This PR introduces a new CUDA compiler handling, following the C++ compiler checks.

Changes:
* `_CCCL_CUDA_COMPILER_*` now expands to the CUDA compiler version (does what `_CCCL_CUDACC_VER` did)
* the current CUDA compiler should be checked using `_CCCL_CUDA_COMPILER(_name [, _op, _maj [, min]])` macro
* the old `_CCCL_CUDA_COMPILER` is replaced with `_CCCL_HAS_CUDA_COMPILER` flag
* the `_CCCL_CUDACC_*` are replaced by `_CCCL_CUDA_COMPILER_*` to match the naming
* `_CCCL_CUDA_COMPILATION` flag can be used to check if CUDA source file is being compiled
* `_CCCL_DEVICE_COMPILATION` flag can be used to check if the device code is being compiled (should replace `defined(__CUDA_ARCH__)`
* the `_CCCL_PTX_ARCH` should be valid for compilation with NVHPC compiler now

There are still many places where the CUDA compiler version is meant to be NVCC only and should be fixed.

### CODE REVIEW

**Comment by bernhardmgruber** [comment]
Please properly rebase this PR onto `main`. I see some changes in the diff that were merged recently.

File: c2h/include/c2h/generators.h [hunk 1]
```diff
@@ -45,12 +45,12 @@ _CCCL_DIAG_SUPPRESS_CLANG("-Wunused-function")
 #    include <cuda_bf16.h>
 _CCCL_DIAG_POP
 
-#    if _CCCL_CUDACC_AT_LEAST(11, 8)
+#    if _CCCL_CUDA_COMPILER_AT_LEAST(11, 8)
```
Comment (miscco) [inline c2h/include/c2h/generators.h:48]: We cannot change that macro, This relates to the `__CUDACC__` macro, aka the version of the cuda headers used 

This is different from the version of e.g clang-cuda and also nvc++ can consume cuda headers from different versions.

So we should leave that macro alone

**Review by miscco — CHANGES_REQUESTED** [review]

File: c2h/include/c2h/generators.h (hunk 1 above)
Comment (miscco) [reply c2h/include/c2h/generators.h:48]: @bernhardmgruber @gevtushenko Do you have a opinions here?

Comment (bernhardmgruber) [reply c2h/include/c2h/generators.h:48]: @miscco it seems you already have a far better understanding what `__CUDACC__` means, so I leave this decision to you.

I agree that if `__CUDACC__` does not designate a CUDA compiler version, then renaming `_CCCL_CUDACC_AT_LEAST` to `_CCCL_CUDA_COMPILER_AT_LEAST` is misleading.

**Comment by davebayer** [comment]
The PR is replaced by #3057.

**Pull request closed without merging**